### PR TITLE
Record failed maintenance operation in shoot event

### LINKED ISSUE #6851: No event or Shoot status information for a failed maintenance operation
**How to categorize this issue?**

/area quality control-plane
/kind bug

**What happened**:
We have API changes that prevent upgrading from one K8s to the next one. For example automatic update of Shoot with `spec.kubernetes.kubeAPIServer.enableBasicAuthentication=true` from K8s 1.18 to 1.19 is not possible:
```
Shoot.core.gardener.cloud \"foo\" is invalid: spec.kubernetes.kubeAPIServer.enableBasicAuthentication: Forbidden: basic authentication has been removed in Kubernetes v1.19+
```

Maybe we have many more cases that prevent the maintenance controller to upgrade K8s/OS versions.

Currently such failure is not reported as event to the Shoot or not reflected in the Shoot status (`.status.lastMaintenance`).

**What you expected to happen**:
A failure of maintenance controller to be reported as event and/or reflected in the Shoot status.

**How to reproduce it (as minimally and precisely as possible)**:
See above.

**Anything else we need to know?**:

**Environment**:

- Gardener version: v1.57.1
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### PR DESCRIPTION
**Pull request by acumino**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement

**What this PR does / why we need it**:
This PR record failed maintenance operation error in shoot events. Original motivation is described in https://github.com/gardener/gardener/issues/6851

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener/issues/6851

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/controllermanager/controller/shoot/maintenance/reconciler.go [hunk 1]
```diff
@@ -156,6 +156,8 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, shoot *gard
 	// try to maintain shoot, but don't retry on conflict, because a conflict means that we potentially operated on stale
 	// data (e.g. when calculating the updated k8s version), so rather return error and backoff
 	if err := gardenClient.Update(ctx, shoot); err != nil {
+		r.Recorder.Eventf(shoot, corev1.EventTypeNormal, "ShootMaintenance", "%s",
+			fmt.Sprintf("Shoot Maintenance Failed, Reason: %s.", err))
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/maintenance/reconciler.go]:
```suggestion
		r.Recorder.Eventf(shoot, corev1.EventTypeWarning, "MaintenanceFailed", err.Error())
```

File: pkg/controllermanager/controller/shoot/maintenance/reconciler.go [hunk 2]
```diff
@@ -156,6 +156,7 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, shoot *gard
 	// try to maintain shoot, but don't retry on conflict, because a conflict means that we potentially operated on stale
 	// data (e.g. when calculating the updated k8s version), so rather return error and backoff
 	if err := gardenClient.Update(ctx, shoot); err != nil {
+		r.Recorder.Event(shoot, corev1.EventTypeWarning, "MaintenanceFailed", err.Error())
```
Comment (ialidzhikov) [inline pkg/controllermanager/controller/shoot/maintenance/reconciler.go]: Cosmetic suggestion: for consistency with the other event reasons, you could move the string to be a constant in

https://github.com/gardener/gardener/blob/36a73eecd54cf757dad68784adc578c5aaf6fcbc/pkg/apis/core/v1beta1/types_shoot.go#L1517 https://github.com/gardener/gardener/blob/36a73eecd54cf757dad68784adc578c5aaf6fcbc/pkg/apis/core/v1alpha1/types_shoot.go#L1504 and https://github.com/gardener/gardener/blob/36a73eecd54cf757dad68784adc578c5aaf6fcbc/pkg/apis/core/types_shoot.go#L1210

**Review by ialidzhikov — COMMENTED** [review]
Thanks for the PR! Minor suggestion, otherwise lgtm.

**Review by ialidzhikov — APPROVED** [review]
/lgtm
/approve

**Pull request merged**